frontend-react/src/components/Admin/MessageTesting.tsx

```diff
@@ -0,0 +1,190 @@
+            },
+            { label: FeatureName.MESSAGE_TESTING },
+        ],
+    };
+
+    if (isDisabled) {
+        return <AdminFetchAlert />;
+    }
+    if (isLoading || !currentTestMessages) return <Spinner />;
+
+    const handleSelect = (event: ChangeEvent<HTMLInputElement>) => {
+        setSelectedOption(event.target.value);
+    };
+
+    const handleAddCustomMessage = () => {
+        setSelectedOption(null);
+        setOpenCustomMessage(true);
+    };
+
+    const CustomMessage = () => {
+        const [text, setText] = useState("");
+        const handleTextareaChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
+            setText(event.target.value);
+        };
+        const handleAddCustomMessage = () => {
+            const dateCreated = new Date();
+            setCurrentTestMessages([
+                ...currentTestMessages,
+                {
+                    dateCreated: dateCreated.toString(),
+                    fileName: `Custom message ${customMessageNumber}`,
+                    reportBody: text,
+                },
+            ]);
+            setCustomMessageNumber(customMessageNumber + 1);
+            setText("");
+            setOpenCustomMessage(false);
+        };
+
+        return (
+            <div className="width-full">
+                <p className="text-bold">Enter custom message</p>
+                <p>Custom messages do not save to the bank after you log out.</p>
+                <Textarea
+                    value={text}
+                    onChange={handleTextareaChange}
+                    id="custom-message-text"
+                    name="custom-message-text"
+                    className="width-full maxw-full margin-bottom-205"
+                />
+                <div className="width-full text-right">
+                    <Button
+                        type="button"
+                        outline
+                        onClick={() => {
+                            setOpenCustomMessage(false);
+                        }}
+                    >
+                        Cancel
+                    </Button>
+                    <Button type="button" onClick={handleAddCustomMessage} disabled={text.length === 0}>
+                        Add
+                    </Button>
+                </div>
+            </div>
+        );
+    };
+
+    const RadioField = ({ title, body, index }: { title: string; body: string; index: number }) => {
+        const openTextInNewTab = () => {
+            let formattedContent = body;
+
+            // Check if the content is JSON and format it
+            try {
+                formattedContent = JSON.stringify(JSON.parse(body), null, 2);
+            } catch {
+                formattedContent = body;
+            }
+
+            const blob = new Blob([formattedContent], { type: "text/plain" });
+
+            const url = URL.createObjectURL(blob);
+
+            window.open(url, "_blank");
```

Comment:
Theoretical concern: is this potentially a race condition if we don't wait for the page to finish loading via the window handle returned before releasing the ObjectURL?

Reply:
I looked into that, and that's only a problem with much larger files, usually images, which we'd use an extremely short `setTimeout` to accommodate.